packages/3dom/src/facade/three-js/correlated-scene-graph.ts

```diff
@@ -0,0 +1,211 @@
+import {Group, Material, Mesh, Object3D, Texture} from 'three';
+import {GLTF as ThreeGLTF, GLTFReference} from 'three/examples/jsm/loaders/GLTFLoader.js';
+
+import {GLTF, GLTFElement} from '../../gltf-2.0.js';
+
+export type ThreeSceneObject = Object3D|Material|Texture;
+type ThreeSceneObjectCallback = (a: ThreeSceneObject, b: ThreeSceneObject) =>
+    void;
+
+export interface GLTFElementHandle {
+  key: Exclude<keyof GLTF, 'scene'>;
+  element: GLTFElement;
+}
+
+export type ThreeObjectSet = Set<ThreeSceneObject>;
+
+export type GLTFElementToThreeObjectMap = Map<GLTFElement, ThreeObjectSet>;
+export type ThreeObjectToGLTFElementHandleMap =
+    Map<Object3D|Material|Texture, GLTFReference>;
+
+const $threeGLTF = Symbol('threeGLTF');
+const $gltf = Symbol('gltf');
+const $gltfElementMap = Symbol('gltfElementMap');
+const $threeObjectMap = Symbol('threeObjectMap');
+const $parallelTraverseThreeScene = Symbol('parallelTraverseThreeScene');
+
+const $correlateOriginalThreeGLTF = Symbol('correlateOriginalThreeGLTF');
+const $correlateCloneThreeGLTF = Symbol('correlateCloneThreeGLTF');
+
+/**
+ * The Three.js GLTFLoader provides us with an in-memory representation
+ * of a glTF in terms of Three.js constructs. It also provides us with a copy
+ * of the deserialized glTF without any Three.js decoration, and a mapping of
+ * glTF elements to their corresponding Three.js constructs.
+ *
+ * A CorrelatedSceneGraph exposes a synchronously available mapping of glTF
+ * element references to their corresponding Three.js constructs.
+ */
+export class CorrelatedSceneGraph {
+  /**
+   * Produce a CorrelatedSceneGraph from a naturally generated Three.js GLTF.
+   * Such GLTFs are produced by Three.js' GLTFLoader, and contain cached
+   * details that expedite the correlation step.
+   *
+   * If a CorrelatedSceneGraph is provided as the second argument, re-correlates
+   * a cloned Three.js GLTF with a clone of the glTF hierarchy used to produce
+   * the upstream Three.js GLTF that the clone was created from. The result
+   * CorrelatedSceneGraph is representative of the cloned hierarchy.
+   */
+  static from(
+      threeGLTF: ThreeGLTF,
+      upstreamCorrelatedSceneGraph?: CorrelatedSceneGraph):
+      CorrelatedSceneGraph {
+    if (upstreamCorrelatedSceneGraph != null) {
+      return this[$correlateCloneThreeGLTF](
+          threeGLTF, upstreamCorrelatedSceneGraph);
+    } else {
+      return this[$correlateOriginalThreeGLTF](threeGLTF);
+    }
+  }
+
+  private static[$correlateOriginalThreeGLTF](threeGLTF: ThreeGLTF):
+      CorrelatedSceneGraph {
+    const gltf = threeGLTF.parser.json as GLTF;
+
+    const {associations} = threeGLTF.parser;
+    const gltfElementMap: GLTFElementToThreeObjectMap = new Map();
+
+    // NOTE: IE11 does not have Map iterator methods
```

Comment:
I thought 3DOM is non-IE11? Am I remembering wrong?

Reply:
Some of this code will still be invoked in IE11. The thing that is not known to be IE11 compatible is the worker script.